dev-docs/RFCs/v7.3/data-management-rfc.md

```diff
@@ -0,0 +1,96 @@
+# RFC: Data Management
+
+* **Author**: Tarek Sherif
+* **Date**: July, 2019
+* **Status**: **Draft**
+
+
+## Summary
+
+This RFC specifies subsystem for managing data in luma.gl and creating gl resources from it.
+
+
+## Background
+
+Buffer and texture management in luma.gl is currently done ad hoc manner, with the application required to build these resources from data and to decide when and how they can be re-used. In deck.gl this has lead to the complexity of the `AttributeManager` that creates buffers from input data and manages their lifecycles.
+
+The lack of a global mechanism for managing data resources also has performance implications in that buffers and textures can be allocated to store redundant data. In deck.gl this might occur due to multiple layers being created from the same data (where each layer would allocate its own gl resources) or multiple instances of the same layer each creating their own layer-specific geometry buffers or texture data. The deck.gl `TileLayer` uses a `TileCache` to avoid redundantly creating textures, and a similar mechanism could be applied more generally to avoid allocating textures and buffers if like resources with the same data have already been created anywhere in the application.
+
+
+## Overview
+
+The proposed data manager would allow the application to register arbitrary named data sets, optionally providing accessors to parse particular views of the data. It would provide API methods to create buffers or textures from the
+registered data, returning cached versions if they are available.
+
+
+## Implementation
+
+A `DataManager` class that supports the following methods:
+- `addData(id, data, accessors)`: add a data set to the manager, optionally with accessors to create particular views.
```

Comment:
I feel that this is just moving the complexity from `AttributeManager` to `DataManager`. I'd prefer a minimal class that is just `ResourceManager` and not deal with data creation.

Reply:
The issue is that the caching requires identifying when the same data is being used. If the accessors are too much, I could just make it take a typed array or an image (i.e. all the parsing would be done in deck, and then the `value` arrays would be registered with the `DataManager`).

Reply:
In deck.gl we do update a typed array in-place before updating the buffer. It's a poor indicator for using cached result since the content of the same typed array and the same buffer may be out of sync.

Reply:
Hmmm... maybe it would be better to make the API closer to the `ProgramManager` then. Something like `dm.addBuffer('myBuffer', buffer)`, where the application would be responsible for providing a key that uniquely identifies a buffer. And all the other operations would be based on buffer ids. Might be tricky to do caching for data, but things like `Geometry` attributes would be easy to cache.